import_models_to_verta/mlflow_databricks_bulk_import_example.ipynb

```diff
@@ -0,0 +1,341 @@
+    "                \"mlflow_last_updated_time:\" + str(rm.last_updated_timestamp),\n",
+    "                \"mlflow_tags:\" + \",\".join(rm.tags),\n",
+    "            ],\n",
+    "        )\n",
+    "        model_messages[rm].append(f\"created new registered model in Verta for {rm.name}\")\n",
+    "    except ValueError:\n",
+    "        model_messages[rm].append(f\"a registered model named \\\"{rm.name}\\\" already exists in Verta. Skipping import.\")\n",
+    "        failed_model_imports.append(rm)\n",
+    "        continue\n",
+    "\n",
+    "    try:\n",
+    "        rm_versions = mlflow_client.search_model_versions(f\"name='{rm.name}'\")\n",
+    "        if not rm_versions:\n",
+    "            failed_model_imports.append(rm)\n",
+    "            model_messages[rm].append(f\"unable to find any model versions for {rm.name}.  Skipping import.\")\n",
+    "            continue\n",
+    "    except Exception as err:\n",
+    "        model_messages[rm].append(f\"failed to fetch versions for registered model \\\"{rm.name}\\\". Skipping import. Error: {err}\")\n",
+    "        failed_model_imports.append(rm)\n",
+    "        continue\n",
+    "\n",
+    "    successful_versions = list()\n",
+    "    for version in rm_versions:\n",
+    "        try:\n",
+    "            verta_version = verta_rm.create_version(\n",
+    "                name=str(version.version),\n",
+    "                attrs={\n",
+    "                    \"er_id\": version.run_id,\n",
+    "                    \"mlflow_source\": version.source,\n",
+    "                    \"mlflow_user_id\": version.user_id,\n",
+    "                    \"mlflow_run_link\": version.run_link,\n",
+    "                    \"mlflow_creation_time\": version.creation_timestamp,\n",
+    "                    \"mlflow_last_updated_time\": version.last_updated_timestamp,\n",
+    "                    \"mlflow_status\": version.status,\n",
+    "                    \"mlflow_current_stage\": version.current_stage,\n",
+    "                    \"mlflow_tags\": version.tags,\n",
+    "                },\n",
+    "                labels=[\"mlflow_import\"],\n",
+    "            )\n",
+    "            model_messages[rm].append(f\"successfully created version {version.version}\")\n",
+    "        except Exception as err:\n",
+    "            model_messages[rm].append(f\"failed to create model version in Verta for {rm.name} - version: {version.version} due to {err}\")\n",
+    "            continue\n",
+    "\n",
+    "        # download artifacts\n",
+    "        try:\n",
+    "            outpath = mlflow.artifacts.download_artifacts(run_id=version.run_id)\n",
+    "        except Exception as err:\n",
+    "            model_messages[rm].append(f\"unable to download artifacts from {rm.name} - version run id; {version.run_id}\")\n",
+    "            outpath = None\n",
+    "\n",
+    "        if outpath:\n",
```

Comment:
Instead of having a `try-except` then a separate `if-else`, we can use a `try-except-else`

```python
# download artifacts
try:
    outpath = mlflow.artifacts.download_artifacts(run_id=version.run_id)
except Exception as err:
    model_messages[rm].append(f"unable to download artifacts from {rm.name} - version run id; {version.run_id}")
else:
    for file_name in os.listdir(outpath):
        if file_name not in protected_file_names:
            verta_version.log_artifact(file_name, os.path.join(outpath, file_name))
            model_messages[rm].append(f"artifact logged in Verta: {file_name}")
        else:
            model_messages[rm].append(f"cannot upload artifact {file_name} for {rm.name} as the file name conflicts with a protected file name in Verta")
```

Reply:
I'm not exactly sure what `download_artifacts()` does when there are no artifacts to be found.  The [source code /docs](https://www.mlflow.org/docs/latest/_modules/mlflow/artifacts.html#download_artifacts) don't really say whether they create an empty directory in that case or do something else.  If they do something else, the subsequent call to `os.listdir(output)` would fail with a `FileNotFoundError`. This is part of my "no working setup to test against problem".

Reply:
I can't imagine that a library would just...create an empty directory, or return an invalid `outpath`, if there were no artifacts.

But...I guess our client has done wilder things :P Maybe the separate `if-else` is fine, if it seems like the safest option in lieu of actually running MLflow ourselves.